File: python/perftest.py
import math
import random
from timeit import default_timer as timer


class Vec3:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def sub(self, v):
        return Vec3(self.x - v.x,
                    self.y - v.y,
                    self.z - v.z)

    def dot(self, v):
        return self.x * v.x + self.y * v.y + self.z * v.z

    def cross(self, v):
        return Vec3(self.y * v.z - self.z * v.y,
                    self.z * v.x - self.x * v.z,
                    self.x * v.y - self.y * v.x)

    def length(self):
        return math.sqrt(self.x * self.x +
                         self.y * self.y +
                         self.z * self.z)

    def normalize(self):
        l = self.length()
        return Vec3(self.x / l, self.y / l, self.z / l)


class Ray:
    def __init__(self, orig=None, direction=None):
        self.orig = orig
        self.direction = direction


def ray_triangle_intersect(r, v0, v1, v2):
    v0v1 = v1.sub(v0)
    v0v2 = v2.sub(v0)
    pvec = r.direction.cross(v0v2)

    det = v0v1.dot(pvec)

    if det < 0.000001:
        return float('-inf')

    invDet = 1.0 / det
    tvec = r.orig.sub(v0)
    u = tvec.dot(pvec) * invDet

    if u < 0 or u > 1:
        return float('-inf')

    qvec = tvec.cross(v0v1)
    v = r.direction.dot(qvec) * invDet

    if v < 0 or u + v > 1:
        return float('-inf')

    return v0v2.dot(qvec) * invDet


def random_vertex():
  return Vec3(random.random() * 2.0 - 1.0,
              random.random() * 2.0 - 1.0,
              random.random() * 2.0 - 1.0)


def generate_random_triangles(numTriangles):
    vertices = [None] * numTriangles * 3

    for i in range(0, numTriangles):
        vertices[i*3 + 0] = random_vertex()
        vertices[i*3 + 1] = random_vertex()
        vertices[i*3 + 2] = random_vertex()

    return vertices


def random_sphere():
    r1 = random.random()
    r2 = random.random()
    lat = math.acos(2*r1 - 1) - math.pi/2
    lon = 2*math.pi * r2

    return Vec3(math.cos(lat) * math.cos(lon),
                math.cos(lat) * math.sin(lon),
                math.sin(lat))


NUM_RAYS = 100
NUM_TRIANGLES = 1000 * 1000

random.seed()

vertices = generate_random_triangles(NUM_TRIANGLES)
num_vertices = NUM_TRIANGLES * 3

num_hit = 0
num_miss = 0

r = Ray()

t_start = timer()

for i in range(0, NUM_RAYS):
    r.orig = random_sphere()
    r.direction = random_sphere().sub(r.orig).normalize()

    for j in range(0, int(num_vertices / 3)):
        t = ray_triangle_intersect(r, vertices[j*3 + 0],
                                      vertices[j*3 + 1],
                                      vertices[j*3 + 2])
        if t >= 0:
          num_hit += 1
        else:
          num_miss += 1

t_end = timer()
t_total = t_end - t_start

num_tests = NUM_RAYS * NUM_TRIANGLES
hit_perc  = float(num_hit) / num_tests * 100
miss_perc = float(num_miss) / num_tests * 100
mtests_per_second = float(num_tests) / t_total / 1000000

#print 'Total intersection tests:  %11d' % (num_tests)
#print '  Hits:\t\t\t   %11d (%5.2f%%)' % (num_hit, hit_perc)
#print '  Misses:\t\t   %11d (%5.2f%%)' % (num_miss, miss_perc)
#print
#print '  Total time:\t\t\t  %6.2f seconds' % (t_total)
#print '  Millions of tests per second:\t  %6.2f' % (mtests_per_second)
print(mtests_per_second)
print(t_total)

File: cpp/perftest.cpp
// Compile with: gcc -std=c++11 -lm -O3 -ffast-math -o perftest perftest.cpp

#include <locale.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <time.h>


/* Vector maths */

struct Vec3
{
  float x;
  float y;
  float z;
};

struct Ray
{
  Vec3 orig;
  Vec3 dir;
};

Vec3 sub(Vec3 a, Vec3 b)
{
  return {a.x - b.x, a.y - b.y, a.z - b.z};
}

float dot(Vec3 a, Vec3 b)
{
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

float len(Vec3 v)
{
  return sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
}

Vec3 normalize(Vec3 v)
{
  float l = len(v);
  return { v.x / l, v.y / l, v.z / l };
}

Vec3 cross(Vec3 a, Vec3 b)
{
  return {
    a.y * b.z - a.z * b.y,
    a.z * b.x - a.x * b.z,
    a.x * b.y - a.y * b.x
  };
}


/* Ray-triangle intersection routine */

float rayTriangleIntersect(Ray *r, Vec3 *v0, Vec3 *v1, Vec3 *v2)
{
  Vec3 v0v1 = sub(*v1, *v0);
  Vec3 v0v2 = sub(*v2, *v0);

  Vec3 pvec = cross(r->dir, v0v2);

  float det = dot(v0v1, pvec);

  bool mask1 = (det >= 0.000001);

  float invDet = 1.0 / det;

  Vec3 tvec = sub(r->orig, *v0);

  float u = dot(tvec, pvec) * invDet;

  bool mask2 = (u >= 0) & (u <= 1);

  Vec3 qvec = cross(tvec, v0v1);

  float v = dot(r->dir, qvec) * invDet;

  bool mask3 = (v >= 0) & ((u + v) <= 1);

  bool all_masks = mask1 & mask2 & mask3;

  float result = dot(v0v2, qvec) * invDet;

  return all_masks ? result : -INFINITY;
}


/* Test data generation */

Vec3 *allocTriangles(int numTriangles)
{
  return (Vec3 *) malloc(sizeof(Vec3) * numTriangles * 3);
}

Vec3 randomSphere()
{
  double r1 = (float) rand() / RAND_MAX;
  double r2 = (float) rand() / RAND_MAX;
  double lat = acos(2*r1 - 1) - M_PI/2;
  double lon = 2*M_PI * r2;

  return {
    (float) (cos(lat) * cos(lon)),
    (float) (cos(lat) * sin(lon)),
    (float) sin(lat)
  };
}

Vec3 *generateRandomTriangles(int numTriangles)
{
  Vec3 *vertices = allocTriangles(numTriangles);

  for (int i = 0; i < numTriangles; ++i)
  {
    vertices[i*3 + 0] = randomSphere();
    vertices[i*3 + 1] = randomSphere();
    vertices[i*3 + 2] = randomSphere();
  }

  return vertices;
}


long ellapsedMs(struct timeval t0, struct timeval t1)
{
  return 1000 * (t1.tv_sec  - t0.tv_sec) +
                (t1.tv_usec - t0.tv_usec) / 1000;
}

int main()
{
  const int NUM_RAYS = 1000;
  const int NUM_TRIANGLES = 100 * 1000;

  srand(time(NULL));

  Vec3 *vertices = generateRandomTriangles(NUM_TRIANGLES);
  const int numVertices = NUM_TRIANGLES * 3;

  int numHit = 0;
  int numMiss = 0;

  Ray r;

  struct timeval tStart;
  gettimeofday(&tStart, 0);

  for (int i = 0; i < NUM_RAYS; ++i)
  {
    r.orig = randomSphere();
    Vec3 p1 = randomSphere();
    r.dir  = normalize((sub(p1, r.orig)));

    for (int j = 0; j < numVertices / 3; ++j)
    {
      float t = rayTriangleIntersect(&r,
                                     &vertices[j*3 + 0],
                                     &vertices[j*3 + 1],
                                     &vertices[j*3 + 2]);
      t >= 0 ? ++numHit : ++numMiss;
    }
  }

  struct timeval tEnd;
  gettimeofday(&tEnd, 0);

  double tTotal = (float) ellapsedMs(tStart, tEnd) / 1000.0;

  int numTests = NUM_RAYS * NUM_TRIANGLES;
  float hitPerc  = ((float) numHit  / numTests) * 100.0f;
  float missPerc = ((float) numMiss / numTests) * 100.0f;
  float mTestsPerSecond = (float) numTests / tTotal / 1000000.0f;

  setlocale(LC_NUMERIC, "");

  printf("Total intersection tests:  %'10i\n", numTests);
  printf("  Hits:\t\t\t    %'10i (%5.2f%%)\n", numHit, hitPerc);
  printf("  Misses:\t\t    %'10i (%5.2f%%)\n", numMiss, missPerc);
  printf("\n");
  printf("Total time:\t\t\t%6.2f seconds\n", tTotal);
  printf("Millions of tests per second:\t%6.2f\n", mTestsPerSecond);
}

